research-software-directory/RSD-as-a-service
# 343 research units

# Create research units and invite maintainers

Closes #343 

Changes proposed in this pull request:
* ONLY primary maintainer is able to create new research units (child organisations). Primary maintainer of the parent organisation is always set as the primary maintainer of child organisation.
* Primary maintainer should then invite others to become "ordinary" maintainers of research unit (organisation)
* Ordinary maintainer CANNOT edit research units, he can edit all other parts of the organisation
* Ordinary maintainer CAN invite others to become "ordinary" maintainer of an organisation/research unit

**Note! on the organisation overview only top level organisation is shown (C_1 in the test example)**

**The implementation deviates from the originally agreed in #343 by**:
* "ordinary" maintainer cannot create research units (child organisations)
* "ordinary" cannot edit research units (child organisations)
* none of the maintainer roles (primary/ordinary) is allowed to delete research units (organisations) 

### Assigning user as maintainer of an organisation
- You can add organisation using any user on any of the software or project items. You will not become maintainer of the organisation automatically (this feature is removed time ago). 
- When you first login as one of the users: professor1 (JRB), professor2 (SAW), professor3 (SIN) the account id is written in the rsd database. You need to assign this account id to primary_maintainer column of an organisation (to become primary maintainer). The sql commands you can use are
```sql
-- LOGIN FIRST THEN NOTE ID
select id from account;
-- ASSIGN PRIMARY MAINTAINER to C01 organisation
-- REPLACE id value in the query with the value you get from the first SQL statement
UPDATE organisation 
    SET primary_maintainer=uuid('c3982c07-0b36-4cf7-b13a-2c2b9932789b')
    WHERE name='C_1'
;
```
How to test:
* `docker-compose down --volumes` to remove old
* `docker-compose build` to build
* `docker-compose up` to start
*  login to RSD as professor1 (JRB)
*  create software S_1
   * create organisation C_1 and assign to S_1 (at organisation section of edit software page)
* make JRB primary maintainer of C_1 organisation using sql script provided
* navigate to C_1 organisation, you should be primary maintainer and see +ADD button
    * add research unit (child organisation) like `C_1_1`
       * navigate to this research unit by clicking on item 
       * go to maintainer section and create maintainer invite email/link. Save link somewhere or send email to yourself ;-)
       *  add research unit at this level, like `C_1_1_1`
           *  navigate to this research unit by clicking on item 
           *  go to maintainer section and create maintainer invite email/link. Save link somewhere or send email to yourself ;-)
* logout from RSD
* login as professor2 (SAW) and use first invite link to become maintainer of C_1_1 research unit (child organisation)
    * you are maintainer of child organisation but not the maintainer of parent
    * as "ordinary" maintainer you can still: pin/block software and projects, change settings and invite other users to become maintainers
    * you can navigate through organisation structure using breadcrumbs
 * login as professor3 (SIN) and use second link to become maintainer of C_1_1_1 research unit

### Example "ordinary" maintainer of research unit "C_1_1_1" (third level of organisation tree)
![image](https://user-images.githubusercontent.com/9204081/175914013-6014bb7b-3f92-4c80-8efc-e41dab0126b1.png)

PR Checklist:
